Revise security considerations section. (#533)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

   compute the set of heavy hitters among a set of arbitrary bit strings
   uploaded by Clients. This requires multiple evaluations of the VDAF, the
   results of which reveal information to the Aggregators and Collector beyond
   what follows from the heavy hitters themselves. Or the result of the Prio3Sum
   VDAF could leak information about outlier values. Note that this leakage can
   be mitigated using differential privacy ({{dp}}).
1. On its own, DAP does not defend against Sybil attacks. See {{sybil}} for
   discussion and potential mitigations.

## Threat model

In this section, we enumerate the actors participating in a Distributed
Aggregation Protocol deployment, enumerate their assets (secrets that are
either inherently valuable or which confer some capability that enables further
attack on the system), the capabilities that a malicious or compromised actor
has, and potential mitigations for attacks enabled by those capabilities.

This model assumes that all participants have previously agreed upon and
exchanged all shared parameters over some unspecified secure channel.

### Client/user

#### Assets

> 1. Unsharded measurements. Clients are the only actor that can ever see the

Author: Note to self: move assets to an appendix (unless I figure out a clever way to keep it in the main text).

Author: Looking back to this list of assets, all of the assets are enumerated in the list of attacks. I think it's more useful to specify the assets with the context of how they can be used/compromised in an attack.

> control a subset of of Clients, one of the Aggregators, and the Collector for a

Author: Please note the change here -- I'm not sure it's useful to define an adversary who can arbitrarily control all of the parties. For example, under the old scenario, the attacker knows the Leader's TLS private key as well as both aggregators' HPKE private keys (since it knows the "secret state of any party"), and is in control of the network, and as such can MitM the Client<->Leader connections then decrypt each measurement as it arrives.

We could call this out as another item outside of our threat model, but I think it makes more sense to restrict attacker control to at most one aggregator -- DAP & VDAF provide no guarantees if the aggregators collude.

> 

Author: Sure -- I include "the client can lie, or submit multiple reports, to skew the results" in attacks on robustness. I think this is technically correct; VDAF defines robustness as "An attacker that controls the network and a subset of Clients cannot cause the Collector to compute anything other than the aggregate of the measurements of honest Clients." DAP will cause the Collector to compute the aggregate of the measurements of honest _and dishonest_ Clients.

   example, the Poplar1 VDAF {{!VDAF}} can be used to compute the set of heavy
   hitters among a set of arbitrary bit strings uploaded by Clients. This
   requires multiple evaluations of the VDAF, the results of which reveal
   information to the Aggregators and Collector beyond what follows from the

>    heavy hitters themselves. Or the result of the Prio3Sum VDAF could leak

Reviewer: Not specific to any VDAF. I think Prio3Sum is meant only as an example, but calling it out specifically may make a casual reader think it's more vulnerable than other VDAFs.
```suggestion
   heavy hitters themselves. Or the result could leak
```

   information about outlier values. This leakage can be mitigated using
   differential privacy ({{dp}}).

[[OPEN ISSUE: Should we have authentication in either direction between the

> Helper and the Collector? #155]]

Reviewer: This issue has been closed, so let's delete this paragraph.

> [[OPEN ISSUE: The threat model for Prio --- as it's described in the original

Reviewer: This open issue is redundant and can be removed. We also plan to call out this threat model in the VDAF draft: https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/304

## Sybil attacks {#sybil}

> Several attacks on privacy or robustness involve malicious Clients uploading
> reports that are valid under the chosen VDAF but incorrect. For example, a DAP

Reviewer: ```suggestion
Several attacks on robustness or privacy involve malicious Clients uploading
reports that are valid under the chosen VDAF but incorrect.

For example, a DAP
```

> deployment might be measuring the heights of a human population and configure a
> VDAF to prove that measurements are values in the range of 80-250 cm. A

Reviewer: ```suggestion
deployment might be measuring the heights of a human population and configure a
variant of Prio3 to prove that measurements are values in the range of 80-250 cm. A
```

privacy violation; or the result may have some property that is desirable to the
adversary ("stats poisoning").

Depending on the deployment and the specific threat being mitigated, there are
different ways to address Sybil attacks, such as:

1. Implementing Client authentication, as described in {{client-auth}}, likely

>    paired with rate-limiting uploads from individual Clients.

Reviewer: We should reference the privacy pass draft: https://datatracker.ietf.org/doc/draft-ietf-privacypass-architecture/

It's in WGLC and shouldn't be a blocker for us.

1. Removing Client-specific metadata on individual reports, such as through the
   use of anonymizing proxies in the upload flow, as described in
   {{anon-proxy}}.

> 1. Applying differential privacy protections, as described in {{dp}}.

Reviewer: We don't want to over-promise here. Whether/how DP is used to beat Sybil attacks is not straightforward.
```suggestion
1. Differential privacy ({{dp}}) can help mitigate Sybil attacks to some extent.
```

## Anonymizing proxies {#anon-proxy}

Client reports can contain auxiliary information such as source IP, HTTP user
agent, or Client authentication information (in deployments which use it, see
{{client-auth}}). This metadata can be used by Aggregators to identify
participating Clients or permit some attacks on robustness. This auxiliary
information can be removed by having Clients submit reports to an anonymizing
proxy server which would then use Oblivious HTTP

> {{!I-D.draft-ietf-ohai-ohttp-08}} to forward reports to the DAP Leader. In this

Reviewer: ```suggestion
{{!I-D.draft-ietf-ohai-ohttp-10}} to forward reports to the DAP Leader. In this
```

scenario, Client authentication would be performed by the proxy rather than any
of the participants in the DAP protocol.

## Differential privacy {#dp}

DAP deployments can choose to ensure their aggregate results achieve

> differential privacy [Vad16]. A simple approach would require the Aggregators to

Reviewer: ```suggestion
differential privacy {{Vad16}}. A simple approach would require the Aggregators to
```

add two-sided noise (e.g. sampled from a two-sided geometric distribution) to
aggregate shares. Since each Aggregator is adding noise independently, privacy
can be guaranteed even if all but one of the Aggregators is malicious.

> Differential privacy is a strong privacy definition, and protects users in

Author: Is this sentence accurate/not overstating things? If I know a single measurement-with-noise value, and the distribution used for noise, I can determine a probability distribution for the measurement-without-noise value. My understanding is that noise values are typically "small", so this distribution would determine a small likely range for the true measurement value.

Author: IIUC, I think what I say is true for "Client DP" (noise added by client to original measurement), but not "Aggregator DP" (noise added by aggregator to aggregate share), looking to https://www.ietf.org/archive/id/draft-wang-ppm-differential-privacy-00.html.

extreme circumstances: even if an adversary has prior knowledge of every
measurement in a batch except for one, that one measurement is still formally
protected.

[[TODO: link to the Shan et al. I-D on differential privacy in DAP once it is

> published.]]

Reviewer: I would probably not add this dependency because it creates a cyclic dependency. (One of these documents will become an RFC before the other.)

tasks increases superlinearly as a function of the chosen bit-length, as
multiple rounds of evaluation are needed for each bit of the measurement value,
and each bit of the measurement value requires computation which scales with the
chosen bit-length.

When dealing with variable length measurements (e.g domain names), it is
necessary to pad them to convert into fixed-length measurements. When computing
the heavy hitters from a batch of such measurements, VDAF execution can be
terminated early once once the padding region is reached for a given
measurement. For smaller length measurements, this significantly reduces the
cost of communication between Aggregators and the steps required for the
computation. However, malicious Clients can still generate maximum length
measurements forcing the system to always operate at worst-case performance.

[[TODO: Revisit this paragraph once
https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/273 is resolved.]]

Therefore, care must be taken that a DAP deployment can handle VDAF execution of
all possible measurement values for any tasks that the deployment is configured
for. Otherwise, an attacker may deny service by uploading many expensive

> reports.

Author: Hmm -- while availability/denial-of-service is often considered a security concern, I agree this section doesn't fit in well since all the other sections are more related to the robustness/privacy properties. Moved.

There was already a subsection ("Data Resolution Limitations") in the Operational Considerations section which touches on basically the same ideas; I pulled a little bit of information in from that subsection & eliminated it.

for. Otherwise, an attacker may deny service by uploading many expensive
reports.

## Differential privacy {#dp}
### Task configuration agreement and consistency

Optionally, DAP deployments can choose to ensure their aggregate results achieve
differential privacy [Vad16]. A simple approach would require the Aggregators to
add two-sided noise (e.g. sampled from a two-sided geometric distribution) to
aggregate shares. Since each Aggregator is adding noise independently, privacy
can be guaranteed even if all but one of the Aggregators is malicious.
Differential privacy is a strong privacy definition, and protects users in
extreme circumstances: even if an adversary has prior knowledge of every
measurement in a batch except for one, that one record is still formally
protected.
In order to execute a DAP task, it is necessary for all parties to ensure they
agree on the configuration of the task. However, it is possible for a party to
participate in the execution of DAP without knowing all of the task's
parameters. For example, a Client can upload a report ({{upload-flow}}) without
knowing the minimum batch size that is enforced by the Aggregators during
collection ({{collect-flow}}).

> ## Robustness in the presence of malicious servers

Author: I removed this section as the relevant issue is closed, and also because we now specify to two aggregators so aggregator subset selection no longer makes sense. (I think something like this idea could work if we had 3+ independent DAP aggregator deployments, sent each report to all deployments, and collected from all of them -- but IMO that's outside of the scope of what needs to be mentioned in the DAP spec.)
